### ISSUE TITLE
Manual mesh definition

### ISSUE THREAD

**Issue by filexora** [mfem/mfem#239]
Hi,

I am trying to bridge mfem with my own application, to use it as FEM solver.

This application already have its own implementation of mesh generator that I am trying
to apply to mfem::Mesh through a set of mixins and traits.

Everything worked fine until I reach a particular point in r.h.s definition:

I am trying to solve a linear elasticity problem similar to ex2.cpp.

I get a segmentation fault when trying to assemble the RHS, it looks like:

`...
VectorGridFunctionCoefficient rhs(&my_gridfunction_rhs);
LinearForm b(fespace);
b.AddBoundaryIntegrator(new mfem::VectorBoundaryLFIntegrator(rhs));
b.Assemble() // <= Segmentation Fault
...`

If I just replace my_gridfunction_rhs by a constant coefficient, everything
work fine, elasticity is solved properly:

`...
VectorGridFunctionCoefficient rhs(new ConstantCoefficient(1.0));
LinearForm b(fespace);
b.AddBoundaryIntegrator(new VectorBoundaryLFIntegrator(rhs));
b.Assemble() // <= NO Segmentation Fault
...`

First I thought that maybe there is a bug when using VectorBoundaryLFIntegrator and VectorGridFunctionCoefficient, as there is no example of this kind of use, so I tried to do the same in ex2.cpp, changing

`   VectorArrayCoefficient f(dim);
   for (int i = 0; i < dim-1; i++)
   {
      f.Set(i, new ConstantCoefficient(0.0));
   }
   {
      Vector pull_force(mesh->bdr_attributes.Max());
      pull_force = 0.0;
      pull_force(1) = -1.0e-2;
      f.Set(dim-1, new PWConstCoefficient(pull_force));
   }`

by

` mfem::GridFunction my_vec(fespace);
 my_vec = 1.0;
VectorGridFunctionCoefficient f(&my_vec);
`
It worked properly.

So I guess I have done something bad during mesh generation, or I am making an improper use of the library... It could looks weird: I define my boundary conditions (here traction force) as a nodal vector field that I integrate on the boundary.

What I do to generate the mesh:
   - adding vertices ( Mesh::AddVertex  )
   - adding hexahedron ( Mesh::AddHex )
   - generating Boundary (Mesh::GenerateBoundaryElement )
   - finalizing (Mesh::Finalize )

Did I miss something that can produce this seg fault ?

Thanks in advance for your help !

**Comment by v-dobrev** [comment]
The first thing I'd recommend is to build mfem with debug mode enabled, e.g. if using make:
```sh
make debug -j 4          # serial debug build
make pdebug -j 4         # parallel debug build
```
(See [INSTALL](https://github.com/mfem/mfem/blob/master/INSTALL) for more details.)

Then rebuild your application and run the case that seg-faults. With debug enabled mfem may catch some issues and point you to the place they appear.

If that does not help (i.e. you still get the seg-fault without any new messages), can you run your code in a debugger and post the stack trace when the seg-fault occurs? That may help us figure out where the issue is.

**Comment by CorporalClegg** [comment]
Hi,

I have done similar work for my ph.D. Here are some advices:

- Try to save the Mesh after building it, open it with glvis and check all is ok
- Try to run the same problem as the one described in ex2.cpp, with the same mesh and output the sparse matrix and the rhs.

What does my_gridfunction_rhs looks like ?

**Comment by filexora** [comment]
Thank you for your fast answers !

I am going to run it with debug on Monday and I come back to tell you the result.

CorporalClegg, my_gridfunction_rhs is a GridFunction defined as:
`GridFunction my_gridfunction_rhs( fespace );

// My forces are represented in a vector of dofs and a vector of value, that I apply on my_gridfunction_rhs. schematically it could be

auto it2 = my_value.begin();
for (auto it = my_dofs.begin() ; it != my_dofs.end() ; it++)
{
     my_gridfunction[*it] = *it2;
     it2++;
}
`

I have already checked that dofs are corresponding, no problem from this side.


Thanks for your advices, I will take a look !

**Comment by filexora** [comment]
Ok, I have some trouble understanding what goes wrong !
Debugger show me that rhs assembly process try to access an element out of range. It seems to me that I am doing something wrong while initialising the mesh. 

I have written a minimum (not) working example, it is a mesh made of a unique element.
It crashes when I try to make b->Assemble():

`
#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
   int order = 1;
   bool static_cond = false;
   bool visualization = 1;

   Mesh *mesh = new Mesh(3,8,1,6);
   int dim = 3;

   // Creating a one-element mesh:
   std::vector<double> pt = {0.0,0.0,0.0};
   mesh->AddVertex(pt.data()); 
   pt = {1.0 , 0.0, 0.0}; 
   mesh->AddVertex(pt.data()); 
   pt = {1.0 , 1.0, 0.0}; 
   mesh->AddVertex(pt.data()); 
   pt = {0.0 , 1.0, 0.0}; 
   mesh->AddVertex(pt.data()); 
   pt = {0.0 , 0.0, 1.0}; 
   mesh->AddVertex(pt.data()); 
   pt = {1.0 , 0.0, 1.0}; 
   mesh->AddVertex(pt.data()); 
   pt = {1.0 , 1.0, 1.0}; 
   mesh->AddVertex(pt.data()); 
   pt = {0.0 , 1.0, 1.0}; 
   mesh->AddVertex(pt.data()); 

   std::vector<int> els = {0,1,2,3,4,5,6,7};
   mesh->AddHex(els.data());

   els = {0,3,2,1};
   mesh->AddBdrQuad(els.data());
   els = {0,1,5,4};
   mesh->AddBdrQuad(els.data());
   els = {3,0,4,7};
   mesh->AddBdrQuad(els.data());
   els = {1,2,6,5};
   mesh->AddBdrQuad(els.data());
   els = {6,2,3,7};
   mesh->AddBdrQuad(els.data());
   els = {4,5,6,7};
   mesh->AddBdrQuad(els.data());


   mesh->FinalizeTopology();
   mesh->Finalize();

   {
      ofstream mesh_ofs("mesh.mesh");
      mesh_ofs.precision(8);
      mesh->Print(mesh_ofs);
   }

   FiniteElementCollection *fec;
   FiniteElementSpace *fespace;
   fec = new H1_FECollection(order, dim);
   fespace = new FiniteElementSpace(mesh, fec, dim);

   // manual definition of boundary conditions: nodes 0,1,2,3 are clamped (dofs set to zero)
   Array<int> ess_tdof_list;
   ess_tdof_list.SetSize(12);
   ess_tdof_list[0 ] = 2;
   ess_tdof_list[1 ] = 3;
   ess_tdof_list[2 ] = 6;
   ess_tdof_list[3 ] = 7;
   ess_tdof_list[4 ] =10;
   ess_tdof_list[5 ] =11;
   ess_tdof_list[6 ] =14;
   ess_tdof_list[7 ] =15;
   ess_tdof_list[8 ] =18;
   ess_tdof_list[9 ] =19;
   ess_tdof_list[10] =22;
   ess_tdof_list[11] =23;

   // manual definition of rhs grid function: force along X on face composed of nodes 0-1-4-5
   mfem::GridFunction my_vec(fespace);
   my_vec[0] = 1.0;
   my_vec[1] = 1.0;
   my_vec[4] = 1.0;
   my_vec[5] = 1.0;

   // Then we build a Coefficient from this vec
   VectorGridFunctionCoefficient f(&my_vec);


   LinearForm *b = new LinearForm(fespace);
   b->AddBoundaryIntegrator(new VectorBoundaryLFIntegrator(f));
   b->Assemble();


   GridFunction x(fespace);
   x = 0.0;

   ConstantCoefficient lambda_func(100.0);
   ConstantCoefficient mu_func(30.0);

   BilinearForm *a = new BilinearForm(fespace);
   a->AddDomainIntegrator(new ElasticityIntegrator(lambda_func,mu_func));

   cout << "matrix ... " << flush;
   if (static_cond) { a->EnableStaticCondensation(); }
   a->Assemble();

   SparseMatrix A;
   Vector B, X;
   a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
   cout << "done." << endl;

   cout << "Size of linear system: " << A.Height() << endl;

   GSSmoother M(A);
   PCG(A, M, B, X, 1, 500, 1e-8, 0.0);

   a->RecoverFEMSolution(X, *b, x);

   if (!mesh->NURBSext)
   {
      mesh->SetNodalFESpace(fespace);
   }

   {
      GridFunction *nodes = mesh->GetNodes();
      *nodes += x;
      x *= -1;
      ofstream mesh_ofs("displaced.mesh");
      mesh_ofs.precision(8);
      mesh->Print(mesh_ofs);
      ofstream sol_ofs("sol.gf");
      sol_ofs.precision(8);
      x.Save(sol_ofs);
   }

   delete a;
   delete b;
   if (fec)
   {
      delete fespace;
      delete fec;
   }
   delete mesh;

   return 0;
}`

NOTE: it require -std=c++11 to compile.

**Comment by v-dobrev** [comment]
I think that the problem is that you are trying to use a `VectorGridFunctionCoefficient` with a boundary integrator. This is not supported  - `*GridFunctionCoefficient`s work only with domain integrators (at the moment).

Specifying a Neumann boundary condition using the values of a `GridFunction` is somewhat non-standard - it can be implemented but there maybe an easier way to setup the same problem without using a `GridFunction` and a `VectorGridFunctionCoefficient`. For example, if your Neumann boundary condition is given as an expression of (x,y,z), you can use a `VectorFunctionCoefficient`. If you need to apply the Neumann boundary condition to just a part of the boundary, you can restrict the `VectorFunctionCoefficient` using a `VectorRestrictedCoefficient` - you just need to mark the corresponding boundary elements with a different boundary attribute when you create the mesh.

The only case, I can think of, where you really need to use a `GridFunction` for the r.h.s. is when your forcing term is obtained by solving another PDE. Is that what you need? If that is the case, I can post a solution for that.

**Comment by filexora** [comment]
Hi,

Ok, not implemented.

Yes it is a bit "strange" to use GridFunctionCoefficient for Neumann conditions, I tried because it offers a really generic way to add Neumann conditions from outside mfem library: just summing values at given dofs.

I can find a way with VectorFunctionCoefficient, but I will probably have to bind it with my own BoundaryCondition structure.

I see VectorFunctionCoefficient use a C-style function, I will definitly need to use functor instead, because my function are user-defined, so runtime defined.

I there are any interest in it, maybe I can provide my implementation of VectorFunctorCoefficient once I have finished ?

Thanks a lot for your help: problem solved !
